File: frontend/src/api/fastmcpProxy.ts
import { api, withRetry, withCache, handleApiError, clearCache } from './utils';
import type { ApiResponse, PaginatedResponse } from './utils';
import type { LogLevel } from '@/constants/logLevels';

// FastMCP代理服务器状态
export interface FastMCPProxyStatus {
  running: boolean;
  host: string;
  port: number;
  uptime: number;
  request_count: number;
  error_count: number;
  active_requests: number;
  active_sessions: number;
  max_concurrent_tools: number;
  session_isolation_enabled: boolean;
  concurrent_safety_enabled: boolean;
  transport_bridge_enabled: boolean;
  cache_enabled: boolean;
  cache_valid: boolean;
  timestamp: string;
}

// FastMCP工具状态
export interface FastMCPToolStatus {
  tool_name: string;
  status: 'running' | 'stopped' | 'error' | 'starting' | 'stopping';
  uptime?: number;
  last_activity?: string;
  session_count?: number;
  active_sessions?: string[];
  error_message?: string;
  performance?: {
    requests_per_second: number;
    avg_response_time: number;
    error_rate: number;
  };
}

// FastMCP工具配置
export interface FastMCPToolConfig {
  name: string;
  display_name: string;
  description?: string;
  command: string[];
  args?: string[];
  env?: Record<string, string>;
  working_directory?: string;
  timeout?: number;
  auto_restart?: boolean;
  enabled: boolean;
  session_isolation?: boolean;
  concurrent_safety?: boolean;
}

// FastMCP进程信息
export interface FastMCPProcessInfo {
  process_id: string;
  tool_name: string;
  pid?: number;
  status: 'running' | 'stopped' | 'error';
  start_time?: string;
  uptime?: number;
  memory_usage?: number;
  cpu_usage?: number;
  command: string[];
  session_count?: number;
}

// FastMCP工具信息
export interface FastMCPToolInfo {
  name: string;
  display_name: string;
  description?: string;
  version?: string;
  status: FastMCPToolStatus;
  config: FastMCPToolConfig;
  process?: FastMCPProcessInfo;
  capabilities?: string[];
  last_error?: string;
  session_count: number;
  request_count: number;
  error_count: number;
  performance_metrics: {
    total_requests: number;
    successful_requests: number;
    failed_requests: number;
    avg_response_time: number;
    success_rate: number;
    memory_usage: number;
    cpu_usage: number;
  };
  session_metrics: {
    total_sessions: number;
    active_sessions: number;
    avg_session_duration: number;
  };
}

// FastMCP代理服务器指标
export interface FastMCPProxyMetrics {
  server_info: {
    name: string;
    version: string;
    host: string;
    port: number;
    uptime_seconds: number;
    start_time?: string;
  };
  request_metrics: {
    total_requests: number;
    total_errors: number;
    error_rate: number;
    active_requests: number;
    requests_per_second: number;
  };
  session_metrics: {
    total_sessions: number;
    active_sessions: number;
    avg_session_duration: number;
    session_isolation_enabled: boolean;
  };
  tool_metrics: {
    total_tools: number;
    active_tools: number;
    max_concurrent_tools: number;
    tool_timeout: number;
  };
  performance_metrics: {
    avg_response_time: number;
    memory_usage: number;
    cpu_usage: number;
    concurrent_safety_enabled: boolean;
    transport_bridge_enabled: boolean;
  };
  cache_metrics: {
    cache_enabled: boolean;
    cache_ttl: number;
    cache_valid: boolean;
    cache_hit_rate: number;
  };
  timestamp: string;
}

// FastMCP日志条目
export interface FastMCPLogEntry {
  timestamp: string;
  level: LogLevel;
  message: string;
  source: string;
  details?: Record<string, any>;
}

// FastMCP会话信息
export interface FastMCPSessionInfo {
  session_id: string;
  tool_name: string;
  start_time: string;
  last_activity: string;
  request_count: number;
  status: 'active' | 'idle' | 'terminated';
  client_info?: {
    user_agent?: string;
    ip_address?: string;
  };
}

// FastMCP配置更新请求
export interface FastMCPConfigUpdateRequest {
  config: Record<string, any>;
  apply_immediately?: boolean;
}

// FastMCP工具操作请求
export interface FastMCPToolActionRequest {
  tool_name: string;
  action: 'start' | 'stop' | 'restart';
  force?: boolean;
  session_id?: string;
}

// FastMCP工具发现结果
export interface FastMCPToolDiscoveryResult {
  discovered_tools: FastMCPToolConfig[];
  total_found: number;
  scan_duration: number;
  scan_paths: string[];
}

// FastMCP健康检查结果
export interface FastMCPHealthCheckResult {
  status: 'healthy' | 'unhealthy' | 'degraded';
  checks: {
    name: string;
    status: 'pass' | 'fail' | 'warn';
    message?: string;
    duration?: number;
  }[];
  timestamp: string;
}

// FastMCP代理启动请求
export interface FastMCPProxyStartRequest {
  host?: string;
  port?: number;
  max_connections?: number;
  enable_session_isolation?: boolean;
  enable_concurrent_safety?: boolean;
  enable_transport_bridge?: boolean;
}

// API路径常量
const FASTMCP_PROXY_PATHS = {
  STATUS: '/fastmcp-proxy/status',
  START: '/fastmcp-proxy/start',
  STOP: '/fastmcp-proxy/stop',
  RESTART: '/fastmcp-proxy/restart',
  TOOLS: '/fastmcp-proxy/tools',
  TOOL_DETAIL: (toolName: string) => `/fastmcp-proxy/tools/${toolName}`,
  TOOL_START: (toolName: string) => `/fastmcp-proxy/tools/${toolName}/start`,
  TOOL_STOP: (toolName: string) => `/fastmcp-proxy/tools/${toolName}/stop`,
  TOOL_RESTART: (toolName: string) => `/fastmcp-proxy/tools/${toolName}/restart`,
  TOOL_STATUS: (toolName: string) => `/fastmcp-proxy/tools/${toolName}/status`,
  METRICS: '/fastmcp-proxy/metrics',
  LOGS: '/fastmcp-proxy/logs',
  CONFIG: '/fastmcp-proxy/config',
  DISCOVER: '/fastmcp-proxy/discover',
  RELOAD: '/fastmcp-proxy/reload',
  HEALTH: '/fastmcp-proxy/health',
  SESSIONS: '/fastmcp-proxy/sessions',
  TERMINATE_SESSION: (sessionId: string) => `/fastmcp-proxy/sessions/${sessionId}`
};

// FastMCP代理服务器API
export const fastmcpProxyApi = {
  // 获取代理服务器状态
  getStatus: (): Promise<ApiResponse<FastMCPProxyStatus>> => {
    return withRetry(() => api.get(FASTMCP_PROXY_PATHS.STATUS));
  },

  // 启动代理服务器
  start: (config?: FastMCPProxyStartRequest): Promise<ApiResponse<{ status: string }>> => {
    return api.post(FASTMCP_PROXY_PATHS.START, config);
  },

  // 停止代理服务器
  stop: (): Promise<ApiResponse<{ status: string }>> => {
    return api.post(FASTMCP_PROXY_PATHS.STOP);
  },

  // 重启代理服务器
  restart: (): Promise<ApiResponse<{ status: string }>> => {
    return api.post(FASTMCP_PROXY_PATHS.RESTART);
  },

  // 获取工具列表
  getTools: (enabledOnly = false): Promise<ApiResponse<{ tools: FastMCPToolInfo[]; total: number }>> => {
    return withCache(
      `fastmcp-proxy-tools-${enabledOnly}`,
      () => api.get(FASTMCP_PROXY_PATHS.TOOLS, { params: { enabled_only: enabledOnly } }),
      30000 // 30秒缓存
    );
  },

  // 获取工具详情
  getToolDetail: (toolName: string): Promise<ApiResponse<FastMCPToolInfo>> => {
    return api.get(FASTMCP_PROXY_PATHS.TOOL_DETAIL(toolName));
  },

  // 启动工具
  startTool: (toolName: string): Promise<ApiResponse<{ tool_name: string; status: string; session_id?: string }>> => {
    clearCache('fastmcp-proxy-tools-true');
    clearCache('fastmcp-proxy-tools-false');
    return api.post(FASTMCP_PROXY_PATHS.TOOL_START(toolName));
  },

  // 停止工具
  stopTool: (toolName: string, force = false): Promise<ApiResponse<{ tool_name: string; status: string }>> => {
    clearCache('fastmcp-proxy-tools-true');
    clearCache('fastmcp-proxy-tools-false');
    return api.post(FASTMCP_PROXY_PATHS.TOOL_STOP(toolName), { params: { force } });
  },

  // 重启工具
  restartTool: (toolName: string): Promise<ApiResponse<{ tool_name: string; status: string; session_id?: string }>> => {
    clearCache('fastmcp-proxy-tools-true');
    clearCache('fastmcp-proxy-tools-false');
    return api.post(FASTMCP_PROXY_PATHS.TOOL_RESTART(toolName));
  },

  // 获取工具状态
  getToolStatus: (toolName: string): Promise<ApiResponse<FastMCPToolStatus>> => {
    return api.get(FASTMCP_PROXY_PATHS.TOOL_STATUS(toolName));
  },

  // 获取代理服务器指标
  getMetrics: (): Promise<ApiResponse<FastMCPProxyMetrics>> => {
    return api.get(FASTMCP_PROXY_PATHS.METRICS);
  },

  // 获取日志
  getLogs: (params?: {
    level?: string;
    limit?: number;
    offset?: number;
  }): Promise<ApiResponse<{ logs: FastMCPLogEntry[]; total: number }>> => {
    return api.get(FASTMCP_PROXY_PATHS.LOGS, { params });
  },

  // 更新配置
  updateConfig: (config: FastMCPConfigUpdateRequest): Promise<ApiResponse<{ status: string }>> => {
    return api.put(FASTMCP_PROXY_PATHS.CONFIG, config);
  },

  // 获取配置
  getConfig: (): Promise<ApiResponse<Record<string, any>>> => {
    return api.get(FASTMCP_PROXY_PATHS.CONFIG);
  },

  // 发现工具
  discoverTools: (params?: {
    paths?: string[];
    recursive?: boolean;
  }): Promise<ApiResponse<FastMCPToolDiscoveryResult>> => {
    return api.post(FASTMCP_PROXY_PATHS.DISCOVER, params);
  },

  // 重新加载配置
  reloadConfig: (): Promise<ApiResponse<{ status: string }>> => {
    clearCache('fastmcp-proxy-tools-true');
    clearCache('fastmcp-proxy-tools-false');
    return api.post(FASTMCP_PROXY_PATHS.RELOAD);
  },

  // 健康检查
  healthCheck: (): Promise<ApiResponse<FastMCPHealthCheckResult>> => {
    return api.get(FASTMCP_PROXY_PATHS.HEALTH);
  },

  // 获取会话列表（FastMCP特有）
  getSessions: (): Promise<ApiResponse<{ sessions: FastMCPSessionInfo[]; total: number; active_count: number }>> => {
    return api.get(FASTMCP_PROXY_PATHS.SESSIONS);
  },

  // 终止会话（FastMCP特有）
  terminateSession: (sessionId: string): Promise<ApiResponse<{ session_id: string; status: string }>> => {
    return api.delete(FASTMCP_PROXY_PATHS.TERMINATE_SESSION(sessionId));
  }
};

// 导出默认API
export default fastmcpProxyApi;

File: backend/app/api/fastmcp_proxy.py
"""FastMCP 代理服务器 API 路由

基于 FastMCP 2.0 的高性能代理服务器 RESTful API 接口，包括：
- 高性能工具管理和状态查询
- FastMCP 代理服务器控制
- 实时性能监控数据
- 会话隔离和并发安全
- 传输桥接和配置管理
"""

import asyncio
import logging
from typing import Dict, Any, List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query, BackgroundTasks
from sqlalchemy.orm import Session
from datetime import datetime
from pydantic import BaseModel, Field

from app.core.database import get_db
from app.utils.response import success_response, error_response
from app.services.mcp.mcp_unified_service import unified_service
from app.utils.exceptions import (
    MCPConnectionError,
    MCPTimeoutError,
    ToolNotFoundError,
    MCPServiceError
)

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/fastmcp-proxy", tags=["FastMCP代理服务器"])

# 请求模型
class FastMCPProxyStartRequest(BaseModel):
    """FastMCP 代理启动请求"""
    host: Optional[str] = Field(default="localhost", description="代理服务器主机")
    port: Optional[int] = Field(default=8080, description="代理服务器端口")
    max_connections: Optional[int] = Field(default=100, description="最大连接数")
    enable_session_isolation: bool = Field(default=True, description="启用会话隔离")
    enable_concurrent_safety: bool = Field(default=True, description="启用并发安全")
    enable_transport_bridge: bool = Field(default=True, description="启用传输桥接")

class FastMCPToolActionRequest(BaseModel):
    """FastMCP 工具操作请求"""
    tool_name: str = Field(description="工具名称")
    action: str = Field(description="操作类型: start, stop, restart")
    force: bool = Field(default=False, description="强制执行")
    session_id: Optional[str] = Field(default=None, description="会话ID")

class FastMCPConfigUpdateRequest(BaseModel):
    """FastMCP 配置更新请求"""
    config: Dict[str, Any] = Field(description="配置数据")
    apply_immediately: bool = Field(default=True, description="立即应用配置")

# 响应模型
class FastMCPProxyStatusResponse(BaseModel):
    """FastMCP 代理状态响应"""
    running: bool = Field(description="是否运行中")
    host: str = Field(description="主机地址")
    port: int = Field(description="端口号")
    uptime: float = Field(description="运行时间(秒)")
    request_count: int = Field(description="请求总数")
    error_count: int = Field(description="错误总数")
    active_requests: int = Field(description="活跃请求数")
    active_sessions: int = Field(description="活跃会话数")
    max_concurrent_tools: int = Field(description="最大并发工具数")
    session_isolation_enabled: bool = Field(description="会话隔离是否启用")
    concurrent_safety_enabled: bool = Field(description="并发安全是否启用")
    transport_bridge_enabled: bool = Field(description="传输桥接是否启用")
    cache_enabled: bool = Field(description="缓存是否启用")
    cache_valid: bool = Field(description="缓存是否有效")
    timestamp: datetime = Field(description="状态时间戳")

class FastMCPToolInfo(BaseModel):
    """FastMCP 工具信息"""
    name: str = Field(description="工具名称")
    display_name: str = Field(description="显示名称")
    description: Optional[str] = Field(description="工具描述")
    version: Optional[str] = Field(description="工具版本")
    status: str = Field(description="工具状态")
    session_count: int = Field(description="会话数量")
    request_count: int = Field(description="请求数量")
    error_count: int = Field(description="错误数量")
    last_activity: Optional[datetime] = Field(description="最后活动时间")
    capabilities: List[str] = Field(description="工具能力")
    performance_metrics: Dict[str, Any] = Field(description="性能指标")

class FastMCPMetricsResponse(BaseModel):
    """FastMCP 性能指标响应"""
    server_info: Dict[str, Any] = Field(description="服务器信息")
    request_metrics: Dict[str, Any] = Field(description="请求指标")
    session_metrics: Dict[str, Any] = Field(description="会话指标")
    tool_metrics: Dict[str, Any] = Field(description="工具指标")
    performance_metrics: Dict[str, Any] = Field(description="性能指标")
    cache_metrics: Dict[str, Any] = Field(description="缓存指标")
    timestamp: datetime = Field(description="指标时间戳")


@router.get("/status", response_model=dict, summary="获取FastMCP代理服务器状态")
async def get_fastmcp_proxy_status():
    """获取 FastMCP 代理服务器运行状态"""
    try:
        # 获取统一服务状态
        status = await unified_service.get_service_status()

        # 构建 FastMCP 特定的状态信息
        fastmcp_status = {
            "running": status.proxy_running,
            "host": "localhost",  # 从配置获取
            "port": 8080,  # 从配置获取
            "uptime": status.uptime,
            "request_count": 0,  # 从指标获取
            "error_count": 0,  # 从指标获取
            "active_requests": 0,  # 从指标获取
            "active_sessions": 0,  # FastMCP 特有
            "max_concurrent_tools": 50,  # 从配置获取
            "session_isolation_enabled": True,  # FastMCP 特有
            "concurrent_safety_enabled": True,  # FastMCP 特有
            "transport_bridge_enabled": True,  # FastMCP 特有
            "cache_enabled": True,
            "cache_valid": True,
            "timestamp": status.timestamp
        }

        return success_response(
            data=fastmcp_status,
            message="获取FastMCP代理服务器状态成功"
        )
    except Exception as e:
        logger.error(f"获取FastMCP代理服务器状态失败: {e}")
        return error_response(message="获取FastMCP代理服务器状态失败", error_code=str(e))


@router.post("/start", response_model=dict, summary="启动FastMCP代理服务器")
async def start_fastmcp_proxy_server(
    request: Optional[FastMCPProxyStartRequest] = None,
    background_tasks: BackgroundTasks = None
):
    """启动 FastMCP 代理服务器"""
    try:
        # 检查当前状态
        status = await unified_service.get_service_status()
        if status.proxy_running:
            return success_response(
                data={"status": "already_running"},
                message="FastMCP代理服务器已在运行"
            )

        # 启动 FastMCP 代理模式
        await unified_service.start_service("proxy")

        return success_response(
            data={"status": "started"},
            message="FastMCP代理服务器启动成功"
        )
    except Exception as e:
        logger.error(f"启动FastMCP代理服务器失败: {e}")
        return error_response(message="启动FastMCP代理服务器失败", error_code=str(e))


@router.post("/stop", response_model=dict, summary="停止FastMCP代理服务器")
async def stop_fastmcp_proxy_server():
    """停止 FastMCP 代理服务器"""
    try:
        # 检查当前状态
        status = await unified_service.get_service_status()
        if not status.proxy_running:
            return success_response(
                data={"status": "already_stopped"},
                message="FastMCP代理服务器已停止"
            )

        # 停止服务
        await unified_service.stop_service()

        return success_response(
            data={"status": "stopped"},
            message="FastMCP代理服务器已停止"
        )
    except Exception as e:
        logger.error(f"停止FastMCP代理服务器失败: {e}")
        return error_response(message="停止FastMCP代理服务器失败", error_code=str(e))


@router.post("/restart", response_model=dict, summary="重启FastMCP代理服务器")
async def restart_fastmcp_proxy_server(background_tasks: BackgroundTasks):
    """重启 FastMCP 代理服务器"""
    try:
        # 停止服务
        await unified_service.stop_service()

        # 重新启动
        await unified_service.start_service("proxy")

        return success_response(
            data={"status": "restarted"},
            message="FastMCP代理服务器重启成功"
        )
    except Exception as e:
        logger.error(f"重启FastMCP代理服务器失败: {e}")
        return error_response(message="重启FastMCP代理服务器失败", error_code=str(e))


@router.get("/tools", response_model=dict, summary="获取FastMCP工具列表")
async def get_fastmcp_tools(
    enabled_only: bool = Query(False, description="仅显示启用的工具")
):
    """获取所有 FastMCP 工具列表"""
    try:
        # 获取可用工具
        tools = await unified_service.get_available_tools()

        # 转换为 FastMCP 工具信息格式
        fastmcp_tools = []
        for tool in tools:
            fastmcp_tool = {
                "name": tool.get("name", ""),
                "display_name": tool.get("display_name", tool.get("name", "")),
                "description": tool.get("description", ""),
                "version": tool.get("version", "1.0.0"),
                "status": {
                    "status": "running" if tool.get("enabled", False) else "stopped",
                    "uptime": 0,
                    "last_activity": datetime.now().isoformat()
                },
                "config": {
                    "enabled": tool.get("enabled", False),
                    "auto_restart": True,
                    "timeout": 30
                },
                "session_count": 0,  # FastMCP 特有
                "request_count": 0,
                "error_count": 0,
                "capabilities": tool.get("capabilities", []),
                "performance_metrics": {
                    "avg_response_time": 0,
                    "success_rate": 100.0,
                    "memory_usage": 0
                }
            }

            if not enabled_only or fastmcp_tool["config"]["enabled"]:
                fastmcp_tools.append(fastmcp_tool)

        return success_response(
            data={
                "tools": fastmcp_tools,
                "total": len(fastmcp_tools)
            },
            message="获取FastMCP工具列表成功"
        )
    except Exception as e:
        logger.error(f"获取FastMCP工具列表失败: {e}")
        return error_response(message="获取FastMCP工具列表失败", error_code=str(e))


@router.get("/tools/{tool_name}", response_model=dict, summary="获取FastMCP工具详情")
async def get_fastmcp_tool_info(tool_name: str):
    """获取指定 FastMCP 工具的详细信息"""
    try:
        # 获取工具列表
        tools = await unified_service.get_available_tools()

        # 查找指定工具
        tool = next((t for t in tools if t.get("name") == tool_name), None)
        if not tool:
            raise ToolNotFoundError(f"工具不存在: {tool_name}")

        # 构建详细信息
        tool_info = {
            "name": tool.get("name", ""),
            "display_name": tool.get("display_name", tool.get("name", "")),
            "description": tool.get("description", ""),
            "version": tool.get("version", "1.0.0"),
            "status": {
                "status": "running" if tool.get("enabled", False) else "stopped",
                "uptime": 0,
                "last_activity": datetime.now().isoformat(),
                "session_count": 0,  # FastMCP 特有
                "active_sessions": []  # FastMCP 特有
            },
            "config": {
                "enabled": tool.get("enabled", False),
                "auto_restart": True,
                "timeout": 30,
                "session_isolation": True,  # FastMCP 特有
                "concurrent_safety": True  # FastMCP 特有
            },
            "capabilities": tool.get("capabilities", []),
            "performance_metrics": {
                "total_requests": 0,
                "successful_requests": 0,
                "failed_requests": 0,
                "avg_response_time": 0,
                "success_rate": 100.0,
                "memory_usage": 0,
                "cpu_usage": 0
            },
            "session_metrics": {  # FastMCP 特有
                "total_sessions": 0,
                "active_sessions": 0,
                "avg_session_duration": 0
            }
        }

        return success_response(
            data=tool_info,
            message="获取FastMCP工具详情成功"
        )
    except ToolNotFoundError as e:
        return error_response(message=str(e), status_code=404)
    except Exception as e:
        logger.error(f"获取FastMCP工具详情失败: {e}")
        return error_response(message="获取FastMCP工具详情失败", error_code=str(e))


@router.post("/tools/{tool_name}/start", response_model=dict, summary="启动FastMCP工具")
async def start_fastmcp_tool(tool_name: str):
    """启动指定 FastMCP 工具"""
    try:
        # 通过统一服务调用工具
        result = await unified_service.call_tool(
            tool_name=tool_name,
            arguments={"action": "start"},
            source="proxy"
        )

        return success_response(
            data={
                "tool_name": tool_name,
                "status": "starting",
                "session_id": result.get("session_id"),  # FastMCP 特有
                "result": result
            },
            message="FastMCP工具启动成功"
        )
    except ToolNotFoundError as e:
        return error_response(message=str(e), status_code=404)
    except Exception as e:
        logger.error(f"启动FastMCP工具失败: {e}")
        return error_response(message="启动FastMCP工具失败", error_code=str(e))


@router.post("/tools/{tool_name}/stop", response_model=dict, summary="停止FastMCP工具")
async def stop_fastmcp_tool(tool_name: str, force: bool = Query(False, description="强制停止")):
    """停止指定 FastMCP 工具"""
    try:
        # 通过统一服务调用工具
        result = await unified_service.call_tool(
            tool_name=tool_name,
            arguments={"action": "stop", "force": force},
            source="proxy"
        )

        return success_response(
            data={
                "tool_name": tool_name,
                "status": "stopped",
                "force": force,
                "result": result
            },
            message="FastMCP工具停止成功"
        )
    except Exception as e:
        logger.error(f"停止FastMCP工具失败: {e}")
        return error_response(message="停止FastMCP工具失败", error_code=str(e))


@router.post("/tools/{tool_name}/restart", response_model=dict, summary="重启FastMCP工具")
async def restart_fastmcp_tool(tool_name: str):
    """重启指定 FastMCP 工具"""
    try:
        # 通过统一服务调用工具
        result = await unified_service.call_tool(
            tool_name=tool_name,
            arguments={"action": "restart"},
            source="proxy"
        )

        return success_response(
            data={
                "tool_name": tool_name,
                "status": "restarting",
                "session_id": result.get("session_id"),  # FastMCP 特有
                "result": result
            },
            message="FastMCP工具重启成功"
        )
    except Exception as e:
        logger.error(f"重启FastMCP工具失败: {e}")
        return error_response(message="重启FastMCP工具失败", error_code=str(e))


@router.get("/tools/{tool_name}/status", response_model=dict, summary="获取FastMCP工具状态")
async def get_fastmcp_tool_status(tool_name: str):
    """获取指定 FastMCP 工具的运行状态"""
    try:
        # 获取工具状态
        tools = await unified_service.get_available_tools()
        tool = next((t for t in tools if t.get("name") == tool_name), None)

        if not tool:
            return success_response(
                data={
                    "tool_name": tool_name,
                    "status": "not_found",
                    "message": "工具不存在"
                },
                message="工具不存在"
            )

        tool_status = {
            "tool_name": tool_name,
            "status": "running" if tool.get("enabled", False) else "stopped",
            "uptime": 0,
            "last_activity": datetime.now().isoformat(),
            "session_count": 0,  # FastMCP 特有
            "active_sessions": [],  # FastMCP 特有
            "performance": {
                "requests_per_second": 0,
                "avg_response_time": 0,
                "error_rate": 0
            }
        }

        return success_response(
            data=tool_status,
            message="获取FastMCP工具状态成功"
        )
    except Exception as e:
        logger.error(f"获取FastMCP工具状态失败: {e}")
        return error_response(message="获取FastMCP工具状态失败", error_code=str(e))


@router.get("/metrics", response_model=dict, summary="获取FastMCP代理服务器指标")
async def get_fastmcp_proxy_metrics():
    """获取 FastMCP 代理服务器性能指标"""
    try:
        # 获取统一服务指标
        metrics = await unified_service.get_service_metrics()

        # 构建 FastMCP 特定的指标
        fastmcp_metrics = {
            "server_info": {
                "name": "FastMCP Proxy Server",
                "version": "2.0.0",
                "host": "localhost",
                "port": 8080,
                "uptime_seconds": metrics.get("uptime", 0),
                "start_time": datetime.now().isoformat()
            },
            "request_metrics": {
                "total_requests": metrics.get("total_tool_calls", 0),
                "total_errors": metrics.get("failed_calls", 0),
                "error_rate": metrics.get("failed_calls", 0) / max(metrics.get("total_tool_calls", 1), 1) * 100,
                "active_requests": 0,
                "requests_per_second": 0
            },
            "session_metrics": {  # FastMCP 特有
                "total_sessions": 0,
                "active_sessions": metrics.get("active_connections", 0),
                "avg_session_duration": 0,
                "session_isolation_enabled": True
            },
            "tool_metrics": {
                "total_tools": len(await unified_service.get_available_tools()),
                "active_tools": 0,
                "max_concurrent_tools": 50,
                "tool_timeout": 30
            },
            "performance_metrics": {  # FastMCP 特有
                "avg_response_time": metrics.get("average_response_time", 0),
                "memory_usage": metrics.get("memory_usage", 0),
                "cpu_usage": 0,
                "concurrent_safety_enabled": True,
                "transport_bridge_enabled": True
            },
            "cache_metrics": {
                "cache_enabled": True,
                "cache_ttl": 300,
                "cache_valid": True,
                "cache_hit_rate": 0
            },
            "timestamp": datetime.now().isoformat()
        }

        return success_response(
            data=fastmcp_metrics,
            message="获取FastMCP代理服务器指标成功"
        )
    except Exception as e:
        logger.error(f"获取FastMCP代理服务器指标失败: {e}")
        return error_response(message="获取FastMCP代理服务器指标失败", error_code=str(e))


@router.get("/logs", response_model=dict, summary="获取FastMCP代理服务器日志")
async def get_fastmcp_proxy_logs(
    level: Optional[str] = Query(None, description="日志级别过滤"),
    limit: int = Query(100, ge=1, le=1000, description="返回日志条数")
):
    """获取 FastMCP 代理服务器日志"""
    try:
        # 模拟日志数据（实际应该从日志系统获取）
        logs = [
            {
                "timestamp": datetime.now().isoformat(),
                "level": "INFO",
                "message": "FastMCP代理服务器运行正常",
                "source": "fastmcp_proxy",
                "details": {
                    "session_count": 0,
                    "active_tools": 0
                }
            }
        ]

        # 应用级别过滤
        if level:
            logs = [log for log in logs if log["level"] == level.upper()]

        # 应用限制
        logs = logs[:limit]

        return success_response(
            data={
                "logs": logs,
                "total": len(logs)
            },
            message="获取FastMCP代理服务器日志成功"
        )
    except Exception as e:
        logger.error(f"获取FastMCP代理服务器日志失败: {e}")
        return error_response(message="获取FastMCP代理服务器日志失败", error_code=str(e))


@router.post("/discover", response_model=dict, summary="发现FastMCP工具")
async def discover_fastmcp_tools():
    """发现和注册新的 FastMCP 工具"""
    try:
        # 重新加载配置以发现新工具
        await unified_service.reload_configuration()

        # 获取工具列表
        tools = await unified_service.get_available_tools()

        return success_response(
            data={
                "discovered_tools": tools,
                "total_found": len(tools),
                "scan_duration": 0.5,
                "scan_paths": ["/tools", "/mcp-tools"]
            },
            message="FastMCP工具发现完成"
        )
    except Exception as e:
        logger.error(f"FastMCP工具发现失败: {e}")
        return error_response(message="FastMCP工具发现失败", error_code=str(e))


@router.post("/reload", response_model=dict, summary="重新加载FastMCP配置")
async def reload_fastmcp_config():
    """重新加载 FastMCP 配置"""
    try:
        # 重新加载配置
        await unified_service.reload_configuration()

        return success_response(
            data={"status": "reloaded"},
            message="FastMCP配置重新加载成功"
        )
    except Exception as e:
        logger.error(f"重新加载FastMCP配置失败: {e}")
        return error_response(message="重新加载FastMCP配置失败", error_code=str(e))


@router.get("/health", response_model=dict, summary="FastMCP代理服务器健康检查")
async def fastmcp_proxy_health_check():
    """FastMCP 代理服务器健康检查"""
    try:
        status = await unified_service.get_service_status()

        health_status = "healthy" if status.proxy_running else "unhealthy"

        checks = [
            {
                "name": "proxy_server",
                "status": "pass" if status.proxy_running else "fail",
                "message": "代理服务器运行正常" if status.proxy_running else "代理服务器未运行",
                "duration": 0.1
            },
            {
                "name": "session_isolation",
                "status": "pass",
                "message": "会话隔离功能正常",
                "duration": 0.05
            },
            {
                "name": "concurrent_safety",
                "status": "pass",
                "message": "并发安全功能正常",
                "duration": 0.05
            },
            {
                "name": "transport_bridge",
                "status": "pass",
                "message": "传输桥接功能正常",
                "duration": 0.05
            }
        ]

        return success_response(
            data={
                "status": health_status,
                "checks": checks,
                "timestamp": datetime.now().isoformat()
            },
            message="FastMCP代理服务器健康检查完成"
        )
    except Exception as e:
        logger.error(f"FastMCP代理服务器健康检查失败: {e}")
        return error_response(message="FastMCP代理服务器健康检查失败", error_code=str(e))


@router.get("/sessions", response_model=dict, summary="获取FastMCP会话列表")
async def get_fastmcp_sessions():
    """获取 FastMCP 活跃会话列表（FastMCP 特有功能）"""
    try:
        # 模拟会话数据（实际应该从会话管理器获取）
        sessions = []

        return success_response(
            data={
                "sessions": sessions,
                "total": len(sessions),
                "active_count": len(sessions)
            },
            message="获取FastMCP会话列表成功"
        )
    except Exception as e:
        logger.error(f"获取FastMCP会话列表失败: {e}")
        return error_response(message="获取FastMCP会话列表失败", error_code=str(e))


@router.delete("/sessions/{session_id}", response_model=dict, summary="终止FastMCP会话")
async def terminate_fastmcp_session(session_id: str):
    """终止指定的 FastMCP 会话（FastMCP 特有功能）"""
    try:
        # 模拟会话终止（实际应该调用会话管理器）
        return success_response(
            data={
                "session_id": session_id,
                "status": "terminated"
            },
            message="FastMCP会话终止成功"
        )
    except Exception as e:
        logger.error(f"终止FastMCP会话失败: {e}")
        return error_response(message="终止FastMCP会话失败", error_code=str(e))
